content/blog/2023/2023-04-23-Java-9-module-system.md

```diff
@@ -0,0 +1,169 @@
+   }
+}
+```
+>NOTE: we can use any module name but we should also keep in mind that the name of the module should follow a reverse domain pattern i.e same naming convention as that of packages.
+
+Now to have control over this module we need to create a **module-info.java file** inside it. As we have discussed above, this file is a module descriptor.
+
+```java
+module first.modules{
+   exports com.demo.helloWorld;
+}
+```
+
+## Creating the second module <a id='creating-the-second-module'></a>
+
+This module depends on our first module as this module will use the perfromOperation function inside the HelloWorld class. To use the first module we should import it inside our second module using **“requires”** directive in the module descriptor of the second module.
+
+```java
+module second.modules{
+   requires first.modules;
+}
+```
+Now we’ll create an application that uses the imported module. For that create a package within the **second.modules** named **com.demo.main.**
+
+Inside this package, we’ll create our **Main.java file** that contains the application code.
+
+```java
+package com.demo.main;
+import com.demo.helloWorld.HelloWorld;
+public class Main {
+   public static void main(String[] args) {
+       HelloWorld.performOperation();
+   }
+}
+```
+## Compiling our modules <a id='compiling-our-modules'></a>
+
+To store the compiled modules we’ll create another directory called **compiledDir** inside the root directory. First, we’ll compile the first module using the below command in the terminal.
+```shell
+javac -d compiledDir --module-source-path project-modules/ --module first.modules
+```
+Now, similarly, we’ll compile our second module using the below command in terminal
+
+```shell
+$ javac -d compiledDir --module-source-path project-modules/ --module second.modules
+```
+Now, after running both of the above commands, we can see the compiled version of both of our modules inside our compiledDir directory.
+
+## Running our Modules <a id='running-our-modules'></a>
+
+Now that we have compiled modules with us, let’s run them using the command
+
+```shell
+$ java --module-path compiledDir -m second.modules/com.demo.main.Main
+```
+On executing the above command we’ll get the following output
+
+```shell
+Hello readers, I'm a Java Module!
+```
+
+## Updated Module Descriptor <a id='updated-module-descriptor'></a>
+
+If we can recall, the module descriptor of our second module only contains the first module. But now, if we execute the below command and have a look at the module descriptor of second module, we can see a change.
+```text
+Compiled from "module-info.java"
+module second.modules {  
+  requires java.base;    
+  requires first.modules;
+}
+```
+Our second module now also contains one more module “java.base” that we didn’t add from our side. This java.base module is actually a default module and all modules are linked to it.
+
+## Conclusion <a id='conclusion'></a>
```

Comment:
I would like two additional sections in this article that talk about these topics:

- How can I create a slimmed-down JDK that includes only the modules of the JDK that I need? (this is the most important aspect of JPMS, in my view) And why would I slim down the JDK?
- Why isn't the module system widely used in application development, yet? (is this because tool support isn't good, yet? I haven't seen it in action, yet ...)